### PR TITLE
[SPARK-5388] Provide a stable application submission gateway for standalone cluster mode

### PR DESCRIPTION
**Pull request by andrewor14**
The goal is to provide a stable, REST-based application submission gateway that is not inherently based on Akka, which is unstable across versions. This PR targets standalone cluster mode, but is implemented in a general enough manner that can be potentially extended to other modes in the future. Client mode is currently not included in the changes here because there are many more Akka messages exchanged there.

As of the changes here, the Master will advertise two ports, 7077 and 6066. We need to keep around the old one (7077) for client mode and older versions of Spark submit. However, all new versions of Spark submit will use the REST gateway (6066).

By the way this includes ~700 lines of tests and ~200 lines of license.

### CODE REVIEW

**Comment by vanzin** [comment]
Hi @andrewor14 , I see the bug is targeted at 1.3. This seems like a pretty considerable change to come so code to the branching point. Should we target it at 1.4 instead?

**Comment by andrewor14** [comment]
@vanzin I understand your concern. Although this patch adds many lines, its scope is actually limited only to standalone cluster mode, and the default submit behavior is actually unchanged. The idea is to provide only an alternative to the existing submission gateway, not a replacement. Also, it should be fairly straightforward to test this, as we only need to test one mode. I will be sure to reiterate on review comments promptly so as not to potentially delay the release.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 1]
```diff
@@ -121,6 +122,17 @@ private[spark] class Master(
     throw new SparkException("spark.deploy.defaultCores must be positive")
   }
 
+  // Alternative application submission gateway that is stable across Spark versions
+  private val restServerEnabled = conf.getBoolean("spark.master.rest.enabled", true)
+  private val restServer =
+    if (restServerEnabled) {
+      val port = conf.getInt("spark.master.rest.port", 17077)
```
Comment (mateiz) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Just a nit, but pick something a bit easier to remember, e.g. 7080 or 7078.

File: core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala [hunk 2]
```diff
@@ -148,15 +148,22 @@ private[deploy] object DeployMessages {
...
+      completedApps: Array[ApplicationInfo],
+      activeDrivers: Array[DriverInfo],
+      completedDrivers: Array[DriverInfo],
+      status: MasterState) {
 
     Utils.checkHost(host, "Required hostname")
     assert (port > 0)
 
     def uri = "spark://" + host + ":" + port
+    def stableUri: Option[String] = stablePort.map { p => "spark://" + host + ":" + p }
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala]: I'm not sure where the right place is, but can you put in some docs or comments on `port` vs `stablePort` & `uri` vs `stableUri`?  or should `stableXXX` be renamed to `restXXX`?

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 1 above)
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: should these confs get added to the docs?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 3]
```diff
@@ -71,21 +81,64 @@ object SparkSubmit {
...
 
-    // Values to return
+  /**
+   * Prepare the environment for submitting an application.
+   * This returns a 4-tuple:
+   *   (1) the arguments for the child process,
+   *   (2) a list of classpath entries for the child,
+   *   (3) a list of system properties and env vars, and
+   *   (4) the main class for the child
+   * Exposed for testing.
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: and it also modified the `args` that are passed in, right?
which is a little confusing as a side effect, why not just return a new, modified args?

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: I was worried about port collisions, since the master is already on 7077. Maybe I can find a better number

Comment (mateiz) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: We can use the "find free port" code to increment it if there's a collision, so I'd rather just pick a number that's easy to remember by default. Especially if this might eventually become the main protocol for app submission.

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+  // Maintain the insert order for converting to JSON later
+  private val fields = new java.util.LinkedHashMap[SubmitRestProtocolField, String]
+  val className = Utils.getFormattedClassName(this)
+
+  // Set the action field
+  fields.put(actionField, action.toString)
+
+  /** Return all fields currently set in this message. */
+  def getFields: Map[SubmitRestProtocolField, String] = fields.toMap
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: why not just use POJOs for each of the message, with normal fields, and just let jackson take care of the conversion to & from json?  Eg. `case class SubmitDriverMessage(action: Action, message: Option[String], ...)`.  I don't see the advantage you get from putting everything into `fields` and doing your own manual type checking, maybe I am missing something fundamental?

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: I'd go one step further. Why implement your own REST server instead of using a JAX-RS implementation like Jersey?

That means that not only a lot of the protocol handling is "automagic", but it also means that you can define the protocol using public classes or interfaces and have MIMA check those for backwards compatibility.

**Comment by vanzin** [comment]
Hi @andrewor14,

>  Although this patch adds many lines, its scope is actually limited only to standalone cluster mode, and the default submit behavior is actually unchanged.

While I understand this, and I also think a stable protocol is a good thing, my main concern is that the bug is marked as a release blocker. Even if this is isolated, it's meant to expose a new public API, so it means that is incurs maintenance for every future release, not just 1.3. So it feels a little rushed to try to get this in at this point, when people for sure will have a lot of feedback.

If feels, from the spec, that making sure the akka-level compatility is not broken in 1.3 would cover a lot of ground already.

**Comment by squito** [comment]
I made a comment at one spot in the code, but throughout I find the name "stable" confusing.  It implies the other one is "unstable", and without the context from the JIRA, unstable to me means that it goes down randomly (not that it has binary compatibility issues across versions).  also if we get rid of the old one, then having the only one be called "stable" will be even stranger.  (Especially if somewhere down the road we find that the new one is unstable in some way.)

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala (hunk 4 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: That is potentially a good alternative and it might simplify the code quite significantly. However I am somewhat against using case classes for stable interfaces in general because it doesn't handle adding or rearranging fields elegantly. We've been using case classes for the `SparkListener` interface and have had to deal with nightmares with providing compatibility. That said I will experiment with using jackson's auto conversion.

**Comment by mateiz** [comment]
If you guys have concerns about the protocol, you should comment on it or on the JIRA; the protocol was posted 5 days ago. Really what matters for this release is *defining* a protocol (such that future versions can implement the spec here). It's useful to put this in 1.3 to prevent the lock-in to Akka versions from extending longer. Basically any protocol that depends on Akka today will definitely be broken in future versions, it's just an accident that it hasn't been yet, so it's best give spark-submit a way to submit to future versions of Spark before that so that as many users as possible can use new versions.

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala (hunk 4 above)
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: are you referring to using java serialization on case classes?  Thats totally different from going to & from json.  Even if you need to use some more advanced api from jackson, eg. adding annotations or something, it still seems like it would make the code a lot simpler.

I'm not familiar w/ jersey but sounds like it might be an even better alternative

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: Just to clarify, Jersey and Jackson are not mutually exclusive.

You'd define your API using Jersey (e.g. `class Foo { void submitMyApp(AppDetails app); }` but with JAX-RS annotations) and then register Jackson as the serializer / deserializer, so that `AppDetails` is "magically" converted to and from JSON.

**Comment by pwendell** [comment]
Hey so in case it's not clear, the issue is that because Akka makes no guarantees whatsoever about compatibility, we need to define our own protocol if we want the standalone cluster manager to be able to accept submissions from different versions. YARN already has this property by virtue of the fact that the YARN API is stable.

In terms of reediness for 1.3 - of course it depends on getting enough support to feel that we can merge this. Since this adds a new additive API (it doesn't change the existing akka-based submission) that can be well tested in isolation, it doesn't seem crazy to me to add it. But of course it depends on getting sign off and making sure it's in good shape.

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala (hunk 4 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: No I wasn't referring to java serialization on case classes. I was referring exactly to using jackson with annotations and stuff.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: The problem with magical conversion is long term compatibility. It is too easy to actually cause incompatible changes. We have a lot of painful experience with that in the past.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: BTW that is almost always the problem with these magic conversions (including a lot of scala magic itself). In the short term they typically provide more concise code, but for something that'd require stability across releases, it is very hard to maintain because it is too easy to break the magic or change the magic.

File: core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala (hunk 2 above)
Comment (rxin) [inline core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala]: this is a great idea!

**Comment by rxin** [comment]
Echoing @squito stable is a bit confusing. Why not just rest?

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala (hunk 4 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: We have a product that uses these "magical tools" and has gone through 10 API revisions with no compatibility issues... so, not convinced, sorry. Manually-written code is *much* more prone to compatibility problems.

Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: I'm with @vanzin on this ... yes I'll agree "magical tools" often don't work as simply as advertised, but thats normally because you're trying to do something complicated, and its only more likely your home grown solution will also have an issue.  Its probably worth writing some unit tests to verify things work as expected, but you've got to do that with anything home grown in any case.

Even say it doesn't work well with scala case classes, and we should use plain java classes, thats still a lot better something totally homegrown.

Just as one tiny example of how this is worse than converting normal java objects: because you're trying to do manual type checking at runtime, you've converted a little typo like `message.setField(SUCCESS,"tru")` from a compile time error to a runtime error.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: Is that how we have been doing it in Spark for all json serialization? Are you suggesting we should change all of them?

Comment (rxin) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: BTW I'm not sure why we are arguing about this. The most important part is the protocol itself, less so about how the protocol is implemented. Implementation changes all the time, and can be changed easily.

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: > Is that how we have been doing it in Spark for all json serialization? Are you suggesting we should change all of them?
I'd love to have the existing JSON serializers in Spark use something like Jackson. `JsonProtocol.scala` is an abomination and a maintenance nightmare. (/imho)

BTW, as Matei requested, we should discuss the protocol in the bug. We're reviewing the code here.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: > I'd love to have the existing JSON serializers in Spark use something like Jackson. JsonProtocol.scala is an abomination and a maintenance nightmare. (/imho)

@vanzin I agree 100%. That's why I went with a more automatic approach in constructing the JSON here. Though I'd argue that the biggest reason why it's a maintenance nightmare is because we're exposing the listener events as public case classes, which really don't support adding new fields without breaking compatibility one way or the other.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: By the way as of the latest commit I went with the jackson approach, which simplifies the code considerably as expected since we no longer have to rely on all the scala class hierarchy magic. Please have a look.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 3 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: because `SparkSubmitArguments` as a bunch of vars and cloning it won't be easy

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: I made this 6066. Let me know if you have any objections

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolField.scala [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.rest
+
+/**
+ * A field used in [[SubmitRestProtocolMessage]]s.
+ */
+class SubmitRestProtocolField[T](val name: String) {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolField.scala]: I think you don't need `name` anymore -- you end up needing to repeat the field name a lot, when now jackson is taking care of putting the field name in the json.  Looks like its only used in `assertFieldIsSet`, which is only called from `DriverStatusRequest` -- so you could just pass in a message for that one case and dry up a lot of the code.

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+  def setMessage(s: String): this.type = setField(message, s)
+
+  /**
+   * Serialize the message to JSON.
+   * This also ensures that the message is valid and its fields are in the expected format.
+   */
+  def toJson: String = {
+    validate()
+    val mapper = new ObjectMapper
+    pretty(parse(mapper.writeValueAsString(this)))
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: is the extra `pretty(parse` to get indentation?  You can do that by configuring the jackson mapper with: `mapper.enable(SerializationFeature.INDENT_OUTPUT)`, so you don't need to reparse the json

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolField.scala (hunk 5 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolField.scala]: sounds fair. An earlier commit had exactly what you suggest here actually. I just thought if we wanted to do extra validation and throw a different exception then we could re-use the name here, but this is no longer the case.

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala (hunk 6 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala]: great

File: core/src/main/scala/org/apache/spark/deploy/rest/SubmitDriverRequest.scala [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  def getDriverExtraJavaOptions: String = driverExtraJavaOptions.toString
+  def getDriverExtraClassPath: String = driverExtraClassPath.toString
+  def getDriverExtraLibraryPath: String = driverExtraLibraryPath.toString
+  def getSuperviseDriver: String = superviseDriver.toString
+  def getExecutorMemory: String = executorMemory.toString
+  def getTotalExecutorCores: String = totalExecutorCores.toString
+
+  // Special getters required for JSON serialization
+  @JsonProperty("appArgs")
+  private def getAppArgsJson: String = arrayToJson(getAppArgs)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/rest/SubmitDriverRequest.scala]: I think you're still doing a lot more work to futz with the json than you to.  I dug into a little bit, and I think the problem is that the json4s wrapper around jackson is lacking a lot of features.  And to use jackson well, you need to add in the [scala module](http://search.maven.org/#artifactdetails%7Ccom.fasterxml.jackson.module%7Cjackson-module-scala_2.10%7C2.3.1%7Cbundle) (or just define these simple bean classes in java).  Eg.:

```
class FooBar{
  var blah: String = _
  var ooga: Array[String] = _
  var wakka: Map[String,String] = _
}
val f = new FooBar
f.blah = "hi"
f.ooga = Array("flim", "flam", "floop")
f.wakka = Map("bip" -> "bop", "beep" -> "boop")
val mapper = new ObjectMapper()
mapper.registerModule(com.fasterxml.jackson.module.scala.DefaultScalaModule)
println(mapper.writeValueAsString(f))
```

prints out `{"blah":"hi","ooga":["flim","flam","floop"],"wakka":{"bip":"bop","beep":"boop"}}`
(it throws an exception w/out the scala module)

**Comment by sryza** [comment]
@pwendell @mateiz I realize I'm chiming in late here, but I think the primary concern is that the protocol doesn't satisfy the principle of least astonishment with respect to what people expect from HTTP submission gateways.  Eventually, whether in Spark itself or in some future project built on top, there will probably end up being a standard way of submitting Spark applications through a REST interface.  If there's a remote chance that this protocol might become that in the future, I think it's worthwhile thinking more critically about it and why we're deviating from more standard REST practices.  For example, I don't think Dale's concerns on the JIRA were really addressed in a rigorous way.  If, on the other hand, the idea is that Spark would only ever use this internally, my personal take is that there's no issue with getting it in for 1.3.

**Comment by pwendell** [comment]
@sryza maybe that is the source of some confusion. This is not intended as a third party extension point at this time. This exists because we want some very simple thing that we can support into the future for standalone job submission because there is no stable API there due to a lack of compatibility guarantees by akka RPC. Nonetheless, I still think we should do some of the proposed improvements like using proper GET and POST calls when needed.

Some bigger "design" questions are out of scope for this. The main one is whether we want to standardize our internal RPC's throughout Spark on an interface that is stable going forward (SPARK-5124). And that comes with questions about whether to use some library to generate bindings, etc.

**Comment by andrewor14** [comment]
Thanks everyone for your comments. It seems that the main common concern is that this protocol does not follow conventions expected in other REST protocols. As of the latest commit, I have addressed this (hopefully) by making the following changes. First, the actions now map to the following URLs and request methods:
```
submit - POST /submissions/create (parameters specified in JSON body)
kill - POST /submissions/kill/driver_123
status - GET / submissions/status/driver_123
```
Although the initial suggestion was to use `DELETE` for kill, doing so actually does not reflect a good use of `DELETE` because we're not removing any resources without side effects.

Second, I introduced protocol versioning and included this in the URL. For instance, a complete URL now looks like `http://1.2.3.4:6066/v1/submissions/create`. A different version here means we break backward and/or forward compatibility between the client and the server, which is expected to almost never happen, however, because this is intended to be a stable submission API.

I also spent some time investigating the use of a JAX-RS framework like Jersey but ultimately decided against it because it introduces many dependencies. Also, although I agree that it hides a lot of path handling details and is much nicer in terms of code readability, it doesn't actually save us many lines since much of the server code actually involves Spark-specific details.

Please let me know of any new or old concerns that I have not addressed.

File: core/src/main/scala/org/apache/spark/deploy/LocalSparkCluster.scala [hunk 8]
```diff
@@ -43,9 +47,11 @@ class LocalSparkCluster(numWorkers: Int, coresPerWorker: Int, memoryPerWorker: I
   def start(): Array[String] = {
     logInfo("Starting a local Spark cluster with " + numWorkers + " workers.")
 
+    // Disable REST server on Master in this mode unless otherwise specified
+    val _conf = conf.clone().setIfMissing("spark.master.rest.enabled", "false")
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/LocalSparkCluster.scala]: why go out of our way to disable this in the local cluster?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 9]
```diff
@@ -300,6 +349,22 @@ private[spark] class SparkSubmitArguments(args: Seq[String], env: Map[String, St
         propertiesFile = value
         parse(tail)
 
+      case ("--kill") :: value :: tail =>
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: this method for having actions with a flag is a bit akward... the Hadoop scripts did a better job with this by having the first word be the required action. Perhaps we should do that at some point (this approach leaves open that possibility).

File: core/src/main/scala/org/apache/spark/deploy/rest/StandaloneRestClient.scala [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+ *
+ * The specific request sent to the server depends on the action as follows:
+ *   (1) submit - POST to /submissions/create
+ *   (2) kill - POST /submissions/kill/[submissionId]
+ *   (3) status - GET /submissions/status/[submissionId]
+ *
+ * In the case of (1), parameters are posted in the HTTP body in the form of JSON fields.
+ * Otherwise, the URL fully specifies the intended action of the client.
+ *
+ * Additionally, the base URL includes the version of the protocol. For instance:
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/rest/StandaloneRestClient.scala]: I would say "as of v1, versions have the following semantics".

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala (hunk 9 above)
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Also, this violates the general format for Spark submit:

```
Usage: spark-submit [options] <app jar | python file> [app options]
```

File: core/src/main/scala/org/apache/spark/deploy/master/MasterMessages.scala [hunk 11]
```diff
@@ -36,7 +36,7 @@ private[master] object MasterMessages {
 
   case object CompleteRecovery
 
-  case object RequestWebUIPort
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterMessages.scala]: Is this just changed for clarity? If so, maybe we should avoid changing it since it will break older akka clients.

File: core/src/main/scala/org/apache/spark/deploy/master/ui/MasterPage.scala [hunk 12]
```diff
@@ -73,6 +73,14 @@ private[spark] class MasterPage(parent: MasterWebUI) extends WebUIPage("") {
           <div class="span12">
             <ul class="unstyled">
               <li><strong>URL:</strong> {state.uri}</li>
+              {
+                state.restUri.map { uri =>
+                  <li>
+                    <strong>REST URL:</strong> {uri}
+                    <span class="rest-uri"> (for standalone cluster mode in Spark 1.3+)</span>
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/MasterPage.scala]: I think it's fine to shorten this to say just say "(cluster mode)" only. The fact that the cluster is a standalone mode cluster and that it is running Spark 1.3 is self-evident.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 13]
```diff
@@ -84,21 +94,92 @@ object SparkSubmit {
...
+  }
+
+  /**
+   * Prepare the environment for submitting an application.
+   * This returns a 4-tuple:
+   *   (1) the arguments for the child process,
+   *   (2) a list of classpath entries for the child,
+   *   (3) a list of system properties and env vars, and
+   *   (4) the main class for the child
+   * In standalone cluster mode, this mutates the original arguments passed in.
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: This also mutates the argumnets for python programs, right?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 14]
```diff
@@ -372,12 +453,28 @@ object SparkSubmit {
       sysProps("spark.submit.pyFiles") = formattedPyFiles
     }
 
+    // NOTE: If we are using the REST gateway, we will use the original arguments directly.
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: I'm very confused by this comment. At this point, we don't actually know whether it will use REST or not, correct?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 15]
```diff
@@ -372,12 +453,28 @@ object SparkSubmit {
...
+    // Since we mutate the values of some configs in this method, we must update the
+    // corresponding fields in the original SparkSubmitArguments to reflect these changes.
+    if (args.isStandaloneCluster) {
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: What about moving this logic outside of this prepare environment function, and just deal with it when you know you are about to do a rest submission. There you have access to the sys props and original arguments.

Also, I really wonder if we can simplify this code. It's very complicated to trace down what is going on here.

File: core/src/main/scala/org/apache/spark/deploy/master/ui/MasterPage.scala (hunk 12 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/MasterPage.scala]: yeah good point

File: core/src/main/scala/org/apache/spark/deploy/master/MasterMessages.scala (hunk 11 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterMessages.scala]: this is actually for tests. Note that we already break older akka clients regardless because we need a way to propagate the REST port to the web UI

File: core/src/main/scala/org/apache/spark/deploy/LocalSparkCluster.scala (hunk 8 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/LocalSparkCluster.scala]: otherwise tests with local-cluster fail

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala (hunk 9 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: as discussed offline, we will continue to use the `--kill` semantics here for now.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 14 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: the semantics are fixed in the latest commit as discussed offline

**Comment by JoshRosen** [comment]
@andrewor14 I saw some of these same NPEs in EasyMock in spurious test failure on a different PR, so I filed https://issues.apache.org/jira/browse/SPARK-5626 for it.

**Comment by andrewor14** [comment]
Thanks @JoshRosen good to know I'm not causing them.

**Comment by pwendell** [comment]
@andrewor14 okay I think this time you are causing the test failure :)

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 16]
```diff
@@ -56,8 +55,16 @@ private[spark] class SparkSubmitArguments(args: Seq[String], env: Map[String, St
   var verbose: Boolean = false
   var isPython: Boolean = false
   var pyFiles: String = null
+  var action: SparkSubmitAction = null
   val sparkProperties: HashMap[String, String] = new HashMap[String, String]()
 
+  // Standalone cluster mode only
+  var useRest: Boolean = true
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: this is not actually settable by the user, so for this one it might be good to indicate it's used only internally

**Comment by pwendell** [comment]
I did a quick pass, this is looking good, but there are some comments on the JIRA worth addressing.

**Comment by andrewor14** [comment]
Alright, as of the latest commit I believe I have addressed all comments both on this PR and on the JIRA. This PR is ready to go from my side.

**Comment by pwendell** [comment]
Okay, this LGTM pending tests. I believe all of the feedback has been addressed. Thanks @vanzin for thorough and very helpful feedback on this patch. This protocol looks much better than it did at the beginning.
